File: day-08/handheld_halting.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

const input_file string = "./day-08/input.txt"

type Instruction struct {
	opCode   string
	argument int
}

type Computer struct {
	program []Instruction
	ip      int
	acc     int
}

func (c *Computer) run() bool {
	executed := make(map[int]bool)
	for {
		if c.ip >= len(c.program) {
			return true // finished execution
		}

		if _, found := executed[c.ip]; found {
			return false // In infinite loop
		} else {
			executed[c.ip] = true
		}
		c.tick()
	}
}

func (c *Computer) tick() {
	instruction := c.program[c.ip]
	switch instruction.opCode {
	case "acc":
		c.acc += instruction.argument
	case "jmp":
		c.ip += instruction.argument
		return //skip ip increment
	}
	c.ip++
}

func ReadProgram(input string) []Instruction {
	instructions := []Instruction{}
	lines := strings.Split(strings.TrimSpace(input), "\n")
	for _, line := range lines {
		instruction_parts := strings.Split(line, " ")
		arg, err := strconv.Atoi(instruction_parts[1])
		if err != nil {
			panic(err)
		}
		instruction := Instruction{
			opCode:   instruction_parts[0],
			argument: arg,
		}
		instructions = append(instructions, instruction)
	}
	return instructions
}

func Part1(input string) int {
	instructions := ReadProgram(input)
	computer := Computer{
		program: instructions,
		ip:      0,
		acc:     0,
	}
	computer.run()
	return computer.acc
}

func Part2(input string) int {
	instructions := ReadProgram(input)
	replacer := strings.NewReplacer("jmp", "nop", "nop", "jmp")
	for i := 0; i < len(instructions); i++ {
		lineCopy := make([]Instruction, len(instructions))
		copy(lineCopy, instructions)
		lineCopy[i].opCode = replacer.Replace(lineCopy[i].opCode)
		computer := Computer{
			program: lineCopy,
			ip:      0,
			acc:     0,
		}
		if computer.run() {
			return computer.acc
		}
	}
	return -1
}

func main() {
	b, err := ioutil.ReadFile(input_file)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %d\n", Part2(data))
}

File: day-08/handheld_halting.py
from collections import defaultdict

input_file = 'day-08/input.txt'

class Program:
  def __init__(self, lines):
    self.lines = lines
    self.pc = 0
    self.regs = defaultdict(int)
    self.executed = set()

  def run(self):
    while self.tick():
      pass
    return self.pc >= len(self.lines)

  def tick(self):
    if not (0 <= self.pc < len(self.lines)):
      return False

    # Stop allowing lines to run twice
    if self.pc in self.executed:
      return False
    else:
      self.executed.add(self.pc)

    instruction, parameter = self.lines[self.pc].split(' ', 1)
    parameter = int(parameter)

    if instruction == 'nop':
      pass
    elif instruction == 'acc':
      self.regs['acc'] += parameter
    elif instruction == 'jmp':
      self.pc += parameter
      return True # skip the pc += 1
    self.pc += 1
    return True

def part1(input):
  program = Program([line.split('#')[0] for line in input.splitlines()])
  program.run()
  return program.get('acc')

def part2(input):
  instructions = [line.split('#')[0] for line in input.splitlines()]

  instruction_change = {'jmp': 'nop', 'nop': 'jmp'}
  for i in range(len(instructions)):
    copy = instructions.copy()
    line_operation = copy[i].split(' ')[0]
    if line_operation in instruction_change:
      copy[i] = copy[i].replace(line_operation, instruction_change[line_operation])
    program = Program(copy)
    if program.run():
      return program.get('acc') 

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))